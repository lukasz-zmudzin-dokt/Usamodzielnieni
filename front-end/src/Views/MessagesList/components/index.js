import ChatForm from "./ChatForm";
import MessageItem from "./MessageItem";

export { 
    ChatForm,
    MessageItem
};