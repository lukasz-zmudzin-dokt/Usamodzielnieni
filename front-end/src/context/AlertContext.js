import React, { useState, createContext } from "react";

export const AlertContext = createContext({
  open: false,
  message: "",
  setOpen: () => {},
});

export const AlertProvider = (props) => {
  const [open, setOpen] = useState(false);
  const [message, setMessage] = useState("");
  const [variant, setVariant] = useState("danger") // success
  const data = {
    open,
    message,
    variant,
    changeVisibility: (newOpen) => setOpen(newOpen),
    changeMessage: (newMessage,newVariant = variant) => {
      setOpen(false);
      setMessage(newMessage);
      setVariant(newVariant);
    },
    showAlert:  (newMessage,newVariant) => {
      data.changeMessage(newMessage,newVariant);
      data.changeVisibility(true);
    }
  };
  return <AlertContext.Provider value={data} {...props} />;
};
