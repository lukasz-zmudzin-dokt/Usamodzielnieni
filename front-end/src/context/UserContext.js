import React, { useState } from "react";
import Cookies from "universal-cookie";

export const UserContext = React.createContext({
  type: undefined, // Employer,Standard,Staff
  token: undefined,
  login: () => {},
  logout: () => {}
});
const cookies = new Cookies();

export const UserProvider = props => {
  const [type, setType] = useState(cookies.get("type"));
  const [token, setToken] = useState(cookies.get("token"));
  const user = {
    token,
    type,
    login: (newToken, newType) => {
      cookies.set("token", newToken, { path: "/" });
      cookies.set("type", newType, { path: "/" });
      setToken(newToken);
      setType(newType);
    },
    logout: () => {
      cookies.remove("token", { path: "/" });
      cookies.remove("type", { path: "/" });
      setToken(undefined);
      setType(undefined);
    }
  };
  return <UserContext.Provider value={user} {...props} />;
};
