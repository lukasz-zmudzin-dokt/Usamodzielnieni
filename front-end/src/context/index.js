export { UserContext, UserProvider } from "context/UserContext";
export { AlertContext, AlertProvider } from "context/AlertContext";
export { ChatContext, ChatProvider } from "context/ChatContext";
export {
  NotificationsContext,
  NotificationsProvider,
} from "context/NotificationsContext";
