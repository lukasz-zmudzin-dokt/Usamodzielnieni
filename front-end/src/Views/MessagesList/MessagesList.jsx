import React, {
  useState,
  useEffect,
  useContext,
  useRef,
  useCallback,
} from "react";
import { ListGroup, Container, Card, Button } from "react-bootstrap";
import MessageItem from "./components/MessageItem";
import { ChatForm } from "./components";
import proxy from "config/api";
import { UserContext, AlertContext, ChatContext } from "context";
import { useParams, useHistory } from "react-router-dom";
import { UserPicture } from "components";

const getMessages = async (token, id) => {
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };
  const url = `${proxy.chat}${id}`;
  const response = await fetch(url, { method: "GET", headers });

  if (response.status === 200) {
    return response.json();
  } else {
    throw response.status;
  }
};

const sendMessage = async (
  token,
  id,
  msg,
  data,
  setData,
  alertC,
  socket,
  user,
  messagesEl
) => {
  const message = {
    message: msg,
    recipient: user.data.username,
  };

  socket.current.send(JSON.stringify(message));
  messagesEl.current.scrollTop = messagesEl.current.scrollHeight;

  socket.current.onerror = (e) => {
    alertC.current.showAlert("Wystąpił błąd.");
  };
};

const MessagesList = () => {
  const [data, setData] = useState([]);
  const user = useContext(UserContext);
  const alertC = useRef(useContext(AlertContext));
  const history = useHistory();
  const { id } = useParams();
  const messagesEl = useRef(null);
  const socket = useRef(null);
  const [contact, setContact] = useState({});
  const chatC = useContext(ChatContext);

  const backToChats = () => {
    if (chatC.socket.current.readyState === WebSocket.CLOSED) {
      chatC.socket.current.send(JSON.stringify({ message: "threads" }));
    }
    history.push("/chats");
  };

  const mapRes = useCallback(
    (res) => {
      const array = res.map((item) => {
        const time = new Date(item.timestamp);
        return {
          content: item.message,
          side: item.sender.username === user.data.username ? "right" : "left",
          send: time.toLocaleString(),
          id: item.timestamp,
        };
      });
      return array;
    },
    [user.data.username]
  );

  const mapAnswer = useCallback(
    (answer) => {
      return {
        content: answer.message,
        side: answer.username === user.data.username ? "right" : "left",
        send: answer.timestamp,
        id: answer.timestamp,
      };
    },
    [user.data.username]
  );

  const checkPhoto = useCallback(
    (res) => {
      if (res.first.username === user.data.username) {
        setContact({ data: { ...res.second } });
      } else {
        setContact({ data: { ...res.first } });
      }
    },
    [user.data.username]
  );

  console.log(contact);

  useEffect(() => {
    const loadMessages = async (token, id) => {
      let res;
      try {
        res = await getMessages(token, id);
        console.log(res);
        setData(mapRes(res.messages));
        checkPhoto(res);
        messagesEl.current.scrollTop = messagesEl.current.scrollHeight;
      } catch (e) {
        console.log(e);
        alertC.current.showAlert("Nie udało się załadować wiadomości.");
        res = [];
      }
    };
    loadMessages(user.token, id);
    messagesEl.current.scrollTop = messagesEl.current.scrollHeight;
  }, [checkPhoto, id, mapRes, user.data.username, user.token]);

  useEffect(() => {
    if (socket.current) {
      socket.current.close();
      socket.current = null;
    }
    if (user.token && user.type) {
      const url = proxy.wsChat + id + "/";

      try {
        socket.current = new WebSocket(url, user.token);
        socket.current.onopen = (e) => {
          messagesEl.current.scrollTop = messagesEl.current.scrollHeight;
        };
        socket.current.onmessage = (object) => {
          setData((prevState) => [
            ...prevState,
            mapAnswer(JSON.parse(JSON.parse(object.data))),
          ]);
          if (messagesEl.current) {
            messagesEl.current.scrollTop = messagesEl.current.scrollHeight;
          }
        };
      } catch (e) {
        console.log(e);
      }
    }
  }, [id, mapAnswer, mapRes, user.token, user.type]);

  return (
    <Container className="messagesList">
      <Card className="messagesList__card">
        <Card.Header as="h3" className="messagesList__header">
          <Button
            variant="outline-warning"
            onClick={backToChats}
            className="messagesList__close"
          >
            {"<"}
          </Button>
          <UserPicture user={contact} />
          <span className="ml-2 messagesList__user">{id}</span>
        </Card.Header>
        <Card.Body className="messagesList__body">
          <ListGroup ref={messagesEl} className="messagesList__list">
            {data.map(({ content, send, side, id }) => (
              <MessageItem key={id} content={content} send={send} side={side} />
            ))}
          </ListGroup>
        </Card.Body>
        <ChatForm
          sendMessage={(msg) =>
            sendMessage(
              user.token,
              data.length,
              msg,
              data,
              setData,
              alertC,
              socket,
              contact,
              messagesEl
            )
          }
        />
      </Card>
    </Container>
  );
};

export default MessagesList;
