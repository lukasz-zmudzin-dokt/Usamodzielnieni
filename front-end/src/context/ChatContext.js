import React, {
  createContext,
  useEffect,
  useState,
  useContext,
  useCallback,
} from "react";
import proxy from "config/api";
import { UserContext } from "context";

const getChats = async (token, filters) => {
  let url = `${proxy.chat}?page=${filters.page}&page_size=${filters.pageSize}`;
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status === 200) {
    return response.json().then((chats) => chats);
  } else {
    throw response.status;
  }
};

export const ChatContext = createContext(null);

export const ChatProvider = (props) => {
  const user = useContext(UserContext);
  const [chats, setChats] = useState([]);
  const [isChatsLoading, setIsChatsLoading] = useState(false);
  const [error, setError] = useState(false);
  const [count, setCount] = useState(0);
  const [filters] = useState({
    page: 1,
    pageSize: 10,
  });
  const [activePage, setActivePage] = useState(1);

  useEffect(() => {
    if (user.token) {
      const loadChats = async (token) => {
        setActivePage(1);
        setError(false);
        setIsChatsLoading(true);
        let loadedChats;
        try {
          loadedChats = await getChats(token, filters);
        } catch (e) {
          console.log(e);
          loadedChats = { count: 0, results: [] };
          setError(true);
        }
        setChats(loadedChats.results);
        setCount(loadedChats.count);
        setIsChatsLoading(false);
      };
      loadChats(user.token);
    }
  }, [filters, user.token]);

  const loadMoreMessages = async () => {
    let res;
    try {
      res = await getChats(user.token, { ...filters, page: activePage + 1 });
      setActivePage(activePage + 1);
    } catch (e) {
      setError(true);
      return;
    }

    setChats([...chats, ...res.results]);
  };

  const loadMessages = async (token) => {
    setActivePage(1);
    let res;
    try {
      res = await getChats(token, filters);
    } catch (e) {
      setError(true);
      return;
    }

    setChats(res.results);
  };

  const data = {
    chats,
    error,
    count,
    loadMoreMessages,
    isChatsLoading,
    loadMessages,
  };

  return <ChatContext.Provider value={data} {...props} />;
};
