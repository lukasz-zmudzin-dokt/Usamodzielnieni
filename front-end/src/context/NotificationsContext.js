import React, {
  useState,
  useContext,
  useEffect,
  useRef,
  createContext,
} from "react";
import { UserContext } from "context";
import proxy from "config/api";
import { paths } from "constants/paths";
import { userTypes } from "constants/userTypes";
import { compile } from "path-to-regexp";

const getNotifications = async (token, type, next) => {
  let url = proxy.notification + "list/all" + (next ? `?page=${next}` : "");
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return response
    .json()
    .then((notifications) => mapNotifications(notifications, type));
};

const getNotificationsCount = async (token) => {
  let url = proxy.notification + "count/unread";
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return (await response.json()).unread_count;
};

const markAsRead = async (token) => {
  let url = proxy.notification + "list/all/mark-as-read/";
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "POST", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return;
};

const deleteNotification = async (token, id) => {
  let url = proxy.notification + `delete/${id}/`;
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "DELETE", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return;
};

const deleteNotifications = async (token) => {
  let url = proxy.notification + "list/all/delete/";
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "DELETE", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return;
};

const mapNotifications = (notifications, type) => ({
  notifications: notifications.results.map((notification) =>
    mapNotification(notification, type)
  ),
  next: notifications.next?.replace(/.*page=/, "") || undefined,
});

const mapNotification = (notification, type) => ({
  id: notification.slug,
  path: getPath(notification.app, notification.object_id, type),
  title: notification.text,
  time: new Date(notification.timestamp),
  unread: notification.unread,
});

const getPath = (appName, id, type) => {
  const isStandard = type === userTypes.STANDARD;
  if (appName.match(/^cv/)) {
    return isStandard
      ? compile(paths.MY_CVS)({})
      : compile(paths.CV_APPROVAL)({});
  } else if (appName.match(/^jobs/)) {
    return isStandard
      ? compile(paths.JOB_OFFERS)({})
      : compile(paths.MY_OFFERS)({});
  } else if (appName.match(/^account/)) {
    return compile(paths.USER)({});
  } else if (appName.match(/^blog/)) {
    return compile(paths.BLOG_POST)({ id });
  } else {
    return compile(paths.DASHBOARD)({});
  }
};

export const NotificationsContext = createContext(null);

export const NotificationsProvider = (props) => {
  const [notifications, setNotifications] = useState(null);
  const [count, setCount] = useState(null);
  const [next, setNext] = useState();
  const [error, setError] = useState(false);
  const socket = useRef();

  const user = useContext(UserContext);

  useEffect(() => {
    const loadNotifications = async (token, type) => {
      setError(false);
      let values;
      try {
        values = await Promise.all([
          getNotifications(token, type),
          getNotificationsCount(token),
        ]);
      } catch (e) {
        console.log(e);
        setError(true);
        return;
      }
      const [
        { notifications: loadedNotifications, next: loadedNext },
        loadedCount,
      ] = values;
      setCount(loadedCount);
      setNext(loadedNext);
      setNotifications(loadedNotifications);
    };
    if (user.token && user.type) {
      loadNotifications(user.token, user.type);
    }
  }, [user.token, user.type]);

  useEffect(() => {
    if (socket.current) {
      socket.current.close();
      socket.current = null;
    }
    if (user.token && user.type) {
      const url = proxy.wsNotification + "ws";
      try {
        socket.current = new WebSocket(url, user.token);
        // socket.current.onopen = (e) => console.log("onopen", e);
        socket.current.onmessage = (e) => {
          const newNotification = mapNotification(
            JSON.parse(e.data),
            user.type
          );
          setCount((prev) => prev + 1);
          setNotifications((prev) => [newNotification, ...prev]);
        };
        socket.current.onerror = (e) => {
          console.log(e);
          setError(true);
        };
      } catch (e) {
        console.log(e);
      }
    }
  }, [user.token, user.type]);

  const data = {
    notifications,
    count,
    error,
    next,
    socket,
    markAsRead: async () => {
      setCount(0);
      try {
        await markAsRead(user.token);
      } catch (e) {
        setError(true);
        return;
      }
    },
    deleteNotification: async (id) => {
      const notificationToRemove = notifications.find(
        (notification) => notification.id === id
      );
      if (notificationToRemove?.unread && count > 0) {
        setCount((prev) => prev - 1);
      }
      setNotifications((prev) =>
        prev.filter((notification) => notification.id !== id)
      );
      try {
        await deleteNotification(user.token, id);
      } catch (e) {
        setError(true);
        return;
      }
    },
    deleteNotifications: async () => {
      setCount(0);
      setNotifications([]);
      try {
        await deleteNotifications(user.token);
      } catch (e) {
        setError(true);
        return;
      }
    },
    loadMoreNotifications: async () => {
      setNext(null);
      let res;
      try {
        res = await getNotifications(user.token, user.type, next);
      } catch (e) {
        setError(true);
        return;
      }
      setNext(res.next);
      setNotifications((prev) => [...prev, ...res.notifications]);
    },
  };
  return <NotificationsContext.Provider value={data} {...props} />;
};
