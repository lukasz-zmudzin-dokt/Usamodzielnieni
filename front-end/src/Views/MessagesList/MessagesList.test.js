import React from "react";
import { render, wait, fireEvent } from "@testing-library/react";
import MessagesList from "Views/MessagesList";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import {AlertContext} from 'context';

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    id: "abc",
  }),
}));

const renderWithRouter = (
  ui,
  {
    route = "/chats/12",
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {}
) => {
  return {
    ...render(<Router history={history}>{ui}</Router>),
    history,
  };
};

describe("MessagesList", () => {
  let failFetch = false;
  let apiMessages = [
    {
      content: "b",
      send: "11:55 12.03.2020",
      side: "left",
      id: 0,
    },
    {
      content: "a",
      send: "11:55 12.03.2020",
      side: "right",
      id: 1,
    },
  ];

  const alertC = {
    showAlert: jest.fn()
  }

  beforeAll(() => {
    global.fetch = jest.fn().mockImplementation((input, init) => {
      return new Promise((resolve, reject) => {
        if (failFetch) {
          resolve({ status: 500 });
        }
        switch (init.method) {
          case "GET":
            resolve({ status: 200, json: () => Promise.resolve(apiMessages) });
            break;
          default:
            reject({});
            break;
        }
      });
    });
  });

  beforeEach(() => {
    failFetch = false;
    jest.clearAllMocks();
  });
  it("should match snapshot", async () => {
    const { container } = renderWithRouter(<MessagesList />);

    await wait(() => expect(fetch).toHaveBeenCalled());

    expect(container).toMatchSnapshot();
  });

  it("should renders [] if api fails", async () => {
    failFetch = true;
    const { queryByText } = renderWithRouter(<AlertContext.Provider value={alertC}><MessagesList /></AlertContext.Provider>);

    await wait(() => expect(alertC.showAlert).toHaveBeenCalled());

    expect(alertC.showAlert).toHaveBeenCalledWith("Nie udało się załadować wiadomości.")
    expect(queryByText("b")).not.toBeInTheDocument();
  });

  it("should redirect to /chats if you click close button", async () => {
    const { getByText, history } = renderWithRouter(<MessagesList />);
    expect(history.location.pathname).toEqual("/chats/12");
    fireEvent.click(getByText("<"));
    await wait(() => expect(fetch).toHaveBeenCalled());
    expect(history.location.pathname).toEqual("/chats");
  });
});
