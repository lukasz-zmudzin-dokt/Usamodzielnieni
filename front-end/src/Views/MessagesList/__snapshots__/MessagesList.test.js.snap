// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesList should match snapshot 1`] = `
<div>
  <div
    class="messagesList container"
  >
    <div
      class="messagesList__card card"
    >
      <h3
        class="messagesList__header card-header"
      >
        <button
          class="messagesList__close btn btn-outline-warning"
          type="button"
        >
          &lt;
        </button>
        <div
          class="userPicture__container"
        >
          <div
            class="userPicture userPicture--standard"
          >
            <span>
              ?
            </span>
          </div>
        </div>
        <span
          class="ml-2"
        >
          Piotr Kowalski
        </span>
      </h3>
      <div
        class="messagesList__body card-body"
      >
        <div
          class="messagesList__list list-group"
        >
          <div
            class="messagesList__item messagesList__item--left list-group-item list-group-item-primary"
          >
            b
          </div>
          <div
            class="messagesList__item messagesList__item--right list-group-item list-group-item-info"
          >
            a
          </div>
        </div>
      </div>
      <div>
        ChatForm
      </div>
    </div>
  </div>
</div>
`;

exports[`MessagesList should renders [] if api fails 1`] = `<div />`;
